bindings/azure/openai/openai.go

```diff
@@ -0,0 +1,283 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openai
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/cognitiveservices/azopenai"
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/kit/config"
+	"github.com/dapr/kit/logger"
+)
+
+// List of operations.
+const (
+	CompletionOperation     bindings.OperationKind = "completion"
+	ChatCompletionOperation bindings.OperationKind = "chat-completion"
+
+	APIKey           = "apiKey"
+	DeploymentId     = "deploymentId"
+	APIBase          = "apiBase"
+	MessagesKey      = "messages"
+	Temperature      = "temperature"
+	MaxTokens        = "maxTokens"
+	TopP             = "topP"
+	N                = "n"
+	Stop             = "stop"
+	FrequencyPenalty = "frequencyPenalty"
+	LogitBias        = "logitBias"
+	User             = "user"
+)
+
+// AzOpenAI represents OpenAI output binding.
+type AzOpenAI struct {
+	logger logger.Logger
+	client *azopenai.Client
+}
+
+type openAIMetadata struct {
+	// APIKey is the API key for the OpenAI API.
+	APIKey string `mapstructure:"apiKey"`
+	// DeploymentId is the deployment ID for the OpenAI API.
+	DeploymentId string `mapstructure:"deploymentId"`
+	// APIBase is the base URL for the OpenAI API.
+	APIBase string `mapstructure:"apiBase"`
+}
+
+type ChatSettings struct {
+	Temperature float32 `mapstructure:"temperature"`
+	MaxTokens   int32   `mapstructure:"maxTokens"`
+	TopP        float32 `mapstructure:"topP"`
+	N           int32   `mapstructure:"n"`
+}
+
+// MessageArray type for chat completion API.
+type MessageArray struct {
+	Messages []Message
+}
+
+type Message struct {
+	Role    string
+	Message string
+}
+
+// Prompt type for completion API.
+type Prompt struct {
+	Prompt string
+}
+
+// NewOpenAI returns a new OpenAI output binding.
+func NewOpenAI(logger logger.Logger) bindings.OutputBinding {
+	return &AzOpenAI{logger: logger}
+}
+
+// Init initializes the OpenAI binding.
+func (p *AzOpenAI) Init(ctx context.Context, meta bindings.Metadata) error {
+	m := openAIMetadata{}
+	err := metadata.DecodeMetadata(meta.Properties, &m)
+	if err != nil {
+		return err
+	}
+	if m.APIKey == "" {
+		return fmt.Errorf("required metadata not set: %s", APIKey)
+	}
+	if m.APIBase == "" {
+		return fmt.Errorf("required metadata not set: %s", APIBase)
+	}
+	if m.DeploymentId == "" {
+		return fmt.Errorf("required metadata not set: %s", DeploymentId)
+	}
+
+	keyCredential, err := azopenai.NewKeyCredential(m.APIKey)
+	p.client, err = azopenai.NewClientWithKeyCredential(m.APIBase, keyCredential, m.DeploymentId, nil)
```

Comment:
For all Azure components, we also ask that you please add support for authentication with Azure AD. In the SDK, the client is created with [`azopenai.NewClient`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/cognitiveservices/azopenai#NewClient). You can see in other components how we use Azure AD. For example: https://github.com/dapr/components-contrib/blob/c62e7c9ad79db775e9d01c0562944f80ff74f37a/state/azure/tablestorage/tablestorage.go#L116-L142

Let me know if you need help with this

Reply:
It's working